### PR TITLE
Add do-while. Modify while-stmt and for-stmt for less branching.

### PR DESCRIPTION
**Pull request by HPCguy**
If you see a clean way to merge the While and DoWhile in stmt() similar to how it was done in gen(), let me know.  Extra code is there to make it "ready" to support break and continue statements.

### CODE REVIEW

File: tests/duff.c [hunk 1]
```diff
@@ -0,0 +1,27 @@
+
```
Comment (jserv) [inline tests/duff.c:1]: Remove the first blank line.

File: tests/duff.c [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+void copy(int *to, int *from, int count)
```
Comment (jserv) [inline tests/duff.c:2]: Use `clang-format` to indent `duff.c`

Comment (jserv) [reply tests/duff.c:2]: However, the bootstrapped `amacc` failed to build `duff`.

**Comment by HPCguy** [comment]
Added full support for break continue for all looping constructs.  Added checks for use of break, continue, default, case outside of usable contexts.

File: tests/duff.c (hunk 2 above)
Comment (HPCguy) [reply tests/duff.c]: I don't have clang-format on my system.

Comment (HPCguy) [reply tests/duff.c]: Have you tried it?  I would expect clang-format to mangle that code.

Comment (HPCguy) [reply tests/duff.c]: Have you tried it?  I would expect clang-format to mangle that code.

File: tests/duff.c (hunk 1 above)
Comment (HPCguy) [reply tests/duff.c]: done

**Comment by HPCguy** [comment]
I tried a bunch of little error tests, but did not submit them because they all halt.  Before I added error checking for case, break, and default statements, if I used those statements anywhere outside of pproper context, the code would mysteriously segfault with no message.  Now, the code provides the line number where those statements are being used out of proper context, and exits gracefully.  I vetted the do-while thoroughly.  I performance tested the reduced branching for all loop types, and got a very tiny speedup (2%?) because the A72 Cortex I am using has such good branch prediction.  I tried a couple of random tests on the new break/continue constructs across loops.

Splasky, thank you in advance for reviewing this.

**Review by jserv — CHANGES_REQUESTED** [review]
Add `do{ ... } while (...)` test in the existing `tests/while.c` along with `break` and `continue` statement.

**Comment by splasky** [comment]
Please send the continue statement in the different pull request.
For adding new opcodes, please also updated the document inside docs/IR.md.
Thanks for your support.

**Comment by HPCguy** [comment]
Splasky... just to be clear, there were only 50 net additional lines of code in amacc.c for all the extra functionality provided by this update.  I will do as you ask if I don't hear back in the next 30 minutes, but it seems to be overkill.

**Comment by HPCguy** [comment]
PS  I am not being flippant when I say this, but I am not very good at git.  I have destroyed several repos when trying to back out of commits, and I am not joking.  The way I will proceed is to delete lines in the relevant test files that I changed and commit those line deletions rather than try to roll anything back.  Sorry about those extra intermediate commits if I have to go down that road.

**Comment by jserv** [comment]
> Splasky... just to be clear, there were only 50 net additional lines of code in amacc.c for all the extra functionality provided by this update. I will do as you ask if I don't hear back in the next 30 minutes, but it seems to be overkill.

I agree with @HPCguy . Let's improve this pull request, adding the descriptions about new opcode in IR documentation.

**Comment by HPCguy** [comment]
I still want approval from Splasky, because I do want to make sure we all agree.

**Review by splasky — APPROVED** [review]

**Comment by HPCguy** [comment]
@splasky : There are actually no new opcode changes in this update that would be appropriate for docs/IR.md.  'Continue' and "DoWhile" are "parse tree" features. Please advise.

**Comment by splasky** [comment]
Oh, Yes. That is my mistake. Only the parser tree changed. You should not need to update the document.

**Comment by HPCguy** [comment]
@jserv : I added break/continue tests to both tests/while.c and tests/for.c and they have been committed.  Feel free to merge at your convenience.  And thank you!

PS  The peephole optimizer is something that could be turned into an academic paper, since it is a tiny "superoptimizer" that could be applied to simple compilers written in a first semester compiler class!

**Comment by jserv** [comment]
Thank @HPCguy for the great work!

**Pull request merged**